### ISSUE TITLE
Possible to show date/time stamps on the X-Axis

### ISSUE THREAD

**Issue by titusjan** [silx-kit/silx#1757]
Hello all. 

I am looking around for various plot libraries and am in the process of selecting one for my project. One of my requirements is to plot dates on the X-Axis. This doesn't seem possible in Silx. 

Would it be possible to implement functionality so that dates or times are displayed next to the ticks the X-axis? I am willing to make pull requests (if I choose to use silx for my project).

I see two approaches:

1. Let silx accept numpy arrays of Python [datetime](https://docs.python.org/2/librarydatetime.html#datetime-objects) objects and detect the array dtype when drawing ticks.
2. Add an option to the PlotWidget and use this option to decide if number or dates at the ticks will be drawn.

I think option 1 is preferable. I tried to make a plot with  array of datetime objects as x-coordinates and I get the following error.

```
Traceback (most recent call last):
  File "/Users/kenter/prog/external/silx/bootstrap.py", line 222, in <module>
    run_file(target, argv)
  File "/Users/kenter/prog/external/silx/bootstrap.py", line 104, in run_file
    execfile(filename, module_globals, module_globals)
  File "/Users/kenter/prog/external/silx/bootstrap.py", line 77, in execfile
    exec(code, globals, locals)
  File "/Users/kenter/work/gssw/src/egse/python/gssw/quicklook/trials/silx_test.py", line 89, in <module>
    main()
  File "/Users/kenter/work/gssw/src/egse/python/gssw/quicklook/trials/silx_test.py", line 83, in main
    makePlot1D()
  File "/Users/kenter/work/gssw/src/egse/python/gssw/quicklook/trials/silx_test.py", line 53, in makePlot1D
    curve = plot1D.addCurve(x=x, y=y, legend='curve')  
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/PlotWidget.py", line 708, in addCurve
    self.resetZoom()
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/PlotWidget.py", line 2577, in resetZoom
    self._forceResetZoom(dataMargins=dataMargins)
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/PlotWidget.py", line 2502, in _forceResetZoom
    ranges = self.getDataRange()
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/PlotWidget.py", line 389, in getDataRange
    self._updateDataRange()
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/PlotWidget.py", line 356, in _updateDataRange
    bounds = item.getBounds()
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/items/core.py", line 180, in getBounds
    return self._getBounds()
  File "/Users/kenter/prog/external/silx/build/lib.macosx-10.7-x86_64-3.5/silx/gui/plot/items/core.py", line 874, in _getBounds
    numpy.nanmin(x),
  File "/Users/kenter/miniconda3/envs/gssw/lib/python3.5/site-packages/numpy/lib/nanfunctions.py", line 238, in nanmin
    if np.isnan(res).any():
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

The issue is that many operation (e.g. `nanmin`) do not work on `datetime` objects so you would have to introduce _special cases_ in a lot of places in the code base. I was hoping that objects or [numpy.datetime64 objects](https://docs.scipy.org/doc/numpy-1.14.0/reference/arrays.datetime.html) would work here but this seems to have the same issues (see https://github.com/numpy/numpy/issues/5610)

I think that therefore option two is better to implement.

What are your thoughts on this? Any help would be appreciated.

**Comment by t20100** [comment]
Hi,

There is indeed no support for dates on the plot axes.
This is something we would like to have, but so far we had no immediate need for it and no time to do it, so if you feel like it, don't hesitate to do it! We can discuss it further and help.

At first glance, I would go for your second proposition: having an API to select how to display the ticks.
I see supporting datetime (at least internally) as more complicated to implement and raising a few issues: it might add a specific case to each tool working with the plot (e.g., Fit and ROI widgets); getting plot limits would return different types depending on the content or need conversion; and what should be done when there is a curve with dates and another one with numbers as X.

For your second proposition, an API to select axis tick mode can be added to the Axis object (http://www.silx.org/doc/silx/latest/modules/gui/plot/items.html#silx.gui.plot.items.Axis)
It can allow to toggle dates/number ticks and to set the number to date conversion (time origin and unit (default: epoch and second)).
This way, silx internals would keep using numbers.

This can then be wrapper in a specific 'time series' widget which activates date ticks by default, and methods like `PlotWidget.addCurve` can be updated to support datetime as input by converting them to numbers.

**Comment by titusjan** [comment]
I started working on the second alternative in my silx-fork. Naturally this took longer than expected, and I only work part-time on this project), but I hope to make a PR with a proof of concept on Friday.

**Comment by t20100** [comment]
Great! Don't hesitate to ask for help, we aim at providing documentation for the public API but the internals are a lot less documented.

**Comment by t20100** [comment]
closed by PR #1770
